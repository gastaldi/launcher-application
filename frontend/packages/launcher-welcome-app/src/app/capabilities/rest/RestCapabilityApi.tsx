import { HttpApi } from '../../../shared/utils/HttpApi';

export interface RestCapabilityApi {
  getGreetingAbsoluteUrl(name: string): string;

  doGetGreeting(name: string): Promise<{ content: string, time: number }>;
}

function buildGreetingPath(path: string, name: string) {
  return name ? `${path}?name=${encodeURIComponent(name)}` : path;
}

export const REST_GREETING_PATH = '/api/greeting';

class HttpRestCapabilityApi implements RestCapabilityApi {

  constructor(private readonly httpApi: HttpApi) {
  }

  public async doGetGreeting(name: string): Promise<{ content: string, time: number }> {
    const greetingPath = buildGreetingPath(REST_GREETING_PATH, name);
    const r = await this.httpApi.get<{ content: string; }>(greetingPath);
    return ({content: r.content, time: Date.now()});
  }

  public getGreetingAbsoluteUrl(name: string): string {
    return this.httpApi.getAbsoluteUrl(buildGreetingPath(REST_GREETING_PATH, name));
  }
}

class MockRestCapabilityApi implements RestCapabilityApi {
  public async doGetGreeting(name: string): Promise<{ content: string, time: number }> {
    return {content: `Hello ${(name || 'World')}!`, time: Date.now()};
  }

  public getGreetingAbsoluteUrl(name: string): string {
    return `http://mocked.io${buildGreetingPath(REST_GREETING_PATH, name)}`;
  }
}

export function newMockRestCapabilityApi(): RestCapabilityApi { return new MockRestCapabilityApi(); }

export const mockRestCapabilityApi = newMockRestCapabilityApi();

export function newHttpRestCapabilityApi(httpApi: HttpApi): RestCapabilityApi {
  return new HttpRestCapabilityApi(httpApi);
}
