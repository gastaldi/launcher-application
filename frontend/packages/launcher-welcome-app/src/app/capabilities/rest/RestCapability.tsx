import { Grid, GridItem, TextInput } from '@patternfly/react-core';
import * as React from 'react';
import { HttpRequest, RequestsConsole, useRequestsState } from '../../../shared/components/HttpRequest';
import { RequestTitle } from '../../../shared/components/RequestTitle';
import { SourceMappingLink } from '../../../shared/components/SourceMappingLink';
import CapabilityCard from '../../components/CapabilityCard';
import capabilitiesConfig from '../../config/capabilitiesConfig';
import { mockRestCapabilityApi, REST_GREETING_PATH } from './RestCapabilityApi';

export interface RestCapabilityProps {
  sourceRepository?: {
    url: string;
    provider: string;
  };
  sourceMapping?: {
    greetingEndpoint: string;
  };
}

export const RestCapabilityApiContext = React.createContext(mockRestCapabilityApi);

export function RestCapability(props: RestCapabilityProps) {
  const api = React.useContext(RestCapabilityApiContext);
  const [requests, addRequestEntry] = useRequestsState();
  const [name, setName] = React.useState<string>('');

  const execGet = () => api.doGetGreeting(name);
  return (
    <CapabilityCard module="rest">
      <CapabilityCard.Title>{capabilitiesConfig.rest.icon} {capabilitiesConfig.rest.name}</CapabilityCard.Title>
      <CapabilityCard.Body>
        <Grid>
          <GridItem span={12}>
            HTTP API endpoints expose your application to outside callers.
            Through these, programs may communicate over the network in a language-independent fashion.
            We have created an initial set of endpoints to illustrate how you may accomplish this in your selected runtime.
            By composing together HTTP endpoints and making use of hypermedia and links,
            you may follow these patterns to construct a RESTful architecture.
            </GridItem>
          <CapabilityCard.Separator />
          <GridItem span={12} className="http-request-service">
            <RequestTitle>
              <SourceMappingLink
                sourceRepository={props.sourceRepository}
                name="greetingEndpoint"
                fileRepositoryLocation={props.sourceMapping && props.sourceMapping.greetingEndpoint}
              />
            </RequestTitle>
          </GridItem>
          <HttpRequest
            name="GET Greetings"
            method="GET"
            url={api.getGreetingAbsoluteUrl(name)}
            path={`${REST_GREETING_PATH}?name=`}
            execute={execGet}
            onRequestResult={addRequestEntry}
          >
            <TextInput
              id="http-api-param-name-input"
              aria-label="Greetings name input"
              value={name}
              onChange={e => setName(e)}
              name="name"
              placeholder="World"
              className="http-request-param"
            />
          </HttpRequest>
          <GridItem span={12}>
            <RequestsConsole name="REST" requests={requests} />
          </GridItem>
        </Grid>
      </CapabilityCard.Body>
    </CapabilityCard>
  );
}
