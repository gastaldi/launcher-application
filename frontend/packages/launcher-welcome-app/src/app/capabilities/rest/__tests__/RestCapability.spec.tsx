import * as React from 'react';
import { act, fireEvent, render, cleanup } from 'react-testing-library';
import { RestCapability, RestCapabilityApiContext } from '../RestCapability';
import { newMockRestCapabilityApi } from '../RestCapabilityApi';
import moment from 'moment-timezone';

const extra = {
  sourceMapping: {
    greetingEndpoint: 'src/main/java/org/your/GreetingApi.java',
  },
};

moment.tz.setDefault('UTC');

jest.useFakeTimers();

beforeAll(() => {
  process.env.TZ = 'UTC';
})

afterEach(() => {
  console.log('cleanup()');
  cleanup();
});

describe('<RestCapability />', () => {
  it('check that initial render is correct', () => {
    const component = render(<RestCapability sourceMapping={extra.sourceMapping} />);
    expect(component).toMatchSnapshot();
  });

  it('check that a click on the GET button adds a Hello World message in the console', async () => {
    const api = newMockRestCapabilityApi();
    const result = { content: 'Hello World!', time: 1542793377 };
    const spy = jest.spyOn(api, 'doGetGreeting').mockResolvedValue(result);
    const Wrapper: React.FunctionComponent = (props) => (<RestCapabilityApiContext.Provider value={api}>{props.children}</RestCapabilityApiContext.Provider>)
    const component = render(<RestCapability sourceMapping={extra.sourceMapping} />, { wrapper: Wrapper });
    fireEvent.click(component.getByLabelText('Execute GET Greetings'));
    await act(async () => {
      await spy.mock.results[0];
    });
    expect(spy).toHaveBeenCalledWith('');
    expect(component.getByLabelText(result.content));
    expect(component.asFragment()).toMatchSnapshot();
  });

  it('check that after typing a name, a click on the GET button adds a Hello John message in the console', async () => {
    const api = newMockRestCapabilityApi();
    const result = { content: 'Hello John!', time: 1542793377 };
    const spy = jest.spyOn(api, 'doGetGreeting').mockResolvedValue(result);
    const Wrapper: React.FunctionComponent = (props) => (<RestCapabilityApiContext.Provider value={api}>{props.children}</RestCapabilityApiContext.Provider>)
    const component = render(<RestCapability sourceMapping={extra.sourceMapping} />, { wrapper: Wrapper });
    fireEvent.change(component.getByLabelText('Greetings name input'), { target: { value: 'John' } });
    expect(component.getByLabelText('Greetings name input').getAttribute('value')).toBe('John');
    fireEvent.click(component.getByLabelText('Execute GET Greetings'));
    await act(async () => {
      await spy.mock.results[0];
    });
    expect(spy).toHaveBeenCalledWith('John');
    expect(component.getByLabelText(result.content));
    expect(component.asFragment()).toMatchSnapshot();
  });

});
