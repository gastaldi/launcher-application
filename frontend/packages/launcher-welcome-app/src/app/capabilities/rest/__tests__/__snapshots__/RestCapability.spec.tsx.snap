// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RestCapability /> check that a click on the GET button adds a Hello World message in the console 1`] = `
<DocumentFragment>
  <article
    class="pf-c-card capability-card"
    id="rest-capability"
  >
    <div
      class="pf-c-card__header capability-card-header"
    >
      <svg
        aria-hidden="true"
        class="with-text"
        fill="currentColor"
        height="1em"
        role="img"
        style="vertical-align: -0.125em;"
        viewBox="0 0 384 512"
        width="1em"
      >
        <path
          d="M256 144V32c0-17.673 14.327-32 32-32s32 14.327 32 32v112h-64zm112 16H16c-8.837 0-16 7.163-16 16v32c0 8.837 7.163 16 16 16h16v32c0 77.406 54.969 141.971 128 156.796V512h64v-99.204c73.031-14.825 128-79.39 128-156.796v-32h16c8.837 0 16-7.163 16-16v-32c0-8.837-7.163-16-16-16zm-240-16V32c0-17.673-14.327-32-32-32S64 14.327 64 32v112h64z"
          transform=""
        />
      </svg>
       HTTP APIs
    </div>
    <div
      class="pf-c-card__body capability-card-body"
    >
      <div
        class="pf-l-grid"
      >
        <div
          class="pf-l-grid__item pf-m-12-col"
        >
          HTTP API endpoints expose your application to outside callers. Through these, programs may communicate over the network in a language-independent fashion. We have created an initial set of endpoints to illustrate how you may accomplish this in your selected runtime. By composing together HTTP endpoints and making use of hypermedia and links, you may follow these patterns to construct a RESTful architecture.
        </div>
        <div
          class="capability-card-separator"
        >
          <hr />
        </div>
        <div
          class="pf-l-grid__item pf-m-12-col http-request-service"
        >
          <svg
            aria-hidden="true"
            class="with-text"
            fill="currentColor"
            height="1em"
            role="img"
            style="vertical-align: -0.125em;"
            viewBox="0 0 448 512"
            width="1em"
          >
            <path
              d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34zm192-34l-136-136c-9.4-9.4-24.6-9.4-33.9 0l-22.6 22.6c-9.4 9.4-9.4 24.6 0 33.9l96.4 96.4-96.4 96.4c-9.4 9.4-9.4 24.6 0 33.9l22.6 22.6c9.4 9.4 24.6 9.4 33.9 0l136-136c9.4-9.2 9.4-24.4 0-33.8z"
              transform=""
            />
          </svg>
           
          <span
            title="src/main/java/org/your/GreetingApi.java"
          >
            greetingEndpoint
          </span>
        </div>
        <div
          class="http-request method-get"
        >
          <div
            class="definition"
          >
            <div
              class="http-request-def"
            >
              <span
                class="http-request-method"
              >
                GET
              </span>
               
              <span
                class="http-request-path"
              >
                /api/greeting?name=
              </span>
              <input
                aria-invalid="false"
                aria-label="Greetings name input"
                class="pf-c-form-control http-request-param"
                id="http-api-param-name-input"
                name="name"
                placeholder="World"
                type="text"
                value=""
              />
              <button
                class="pf-c-button pf-m-primary copy-button"
                title="Copy as curl"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  fill="currentColor"
                  height="1em"
                  role="img"
                  style="vertical-align: -0.125em;"
                  viewBox="0 0 384 512"
                  width="1em"
                >
                  <path
                    d="M384 112v352c0 26.51-21.49 48-48 48H48c-26.51 0-48-21.49-48-48V112c0-26.51 21.49-48 48-48h80c0-35.29 28.71-64 64-64s64 28.71 64 64h80c26.51 0 48 21.49 48 48zM192 40c-13.255 0-24 10.745-24 24s10.745 24 24 24 24-10.745 24-24-10.745-24-24-24m96 114v-20a6 6 0 0 0-6-6H102a6 6 0 0 0-6 6v20a6 6 0 0 0 6 6h180a6 6 0 0 0 6-6z"
                    transform=""
                  />
                </svg>
              </button>
            </div>
          </div>
          <div
            class="action"
          >
            <button
              aria-label="Execute GET Greetings"
              class="pf-c-button pf-m-primary http-request-button"
              title="Execute GET Greetings"
              type="button"
            >
              Execute
            </button>
          </div>
        </div>
        <div
          class="pf-l-grid__item pf-m-12-col"
        >
          <div
            aria-label="REST Console"
            class="console"
          >
            <div
              class="console-content expanded"
            >
              <span
                class="placeholder"
              >
                <span
                  class="prefix"
                >
                  &gt;_
                </span>
                  Use the 'Execute' button(s) above to get things started in this console...
              </span>
              <div
                class="console-content-row"
              >
                <div>
                  <span
                    class="prefix"
                  >
                    $
                  </span>
                   
                  <span
                    class="time"
                  >
                    8:33:13 PM
                  </span>
                   
                  <span
                    class="method method-get"
                  >
                    GET
                  </span>
                   
                  <span
                    class="url"
                  >
                    http://mocked.io/api/greeting
                  </span>
                  :
                </div>
                <div
                  aria-label="Hello World!"
                >
                  Hello World!
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </article>
</DocumentFragment>
`;

exports[`<RestCapability /> check that after typing a name, a click on the GET button adds a Hello John message in the console 1`] = `
<DocumentFragment>
  <article
    class="pf-c-card capability-card"
    id="rest-capability"
  >
    <div
      class="pf-c-card__header capability-card-header"
    >
      <svg
        aria-hidden="true"
        class="with-text"
        fill="currentColor"
        height="1em"
        role="img"
        style="vertical-align: -0.125em;"
        viewBox="0 0 384 512"
        width="1em"
      >
        <path
          d="M256 144V32c0-17.673 14.327-32 32-32s32 14.327 32 32v112h-64zm112 16H16c-8.837 0-16 7.163-16 16v32c0 8.837 7.163 16 16 16h16v32c0 77.406 54.969 141.971 128 156.796V512h64v-99.204c73.031-14.825 128-79.39 128-156.796v-32h16c8.837 0 16-7.163 16-16v-32c0-8.837-7.163-16-16-16zm-240-16V32c0-17.673-14.327-32-32-32S64 14.327 64 32v112h64z"
          transform=""
        />
      </svg>
       HTTP APIs
    </div>
    <div
      class="pf-c-card__body capability-card-body"
    >
      <div
        class="pf-l-grid"
      >
        <div
          class="pf-l-grid__item pf-m-12-col"
        >
          HTTP API endpoints expose your application to outside callers. Through these, programs may communicate over the network in a language-independent fashion. We have created an initial set of endpoints to illustrate how you may accomplish this in your selected runtime. By composing together HTTP endpoints and making use of hypermedia and links, you may follow these patterns to construct a RESTful architecture.
        </div>
        <div
          class="capability-card-separator"
        >
          <hr />
        </div>
        <div
          class="pf-l-grid__item pf-m-12-col http-request-service"
        >
          <svg
            aria-hidden="true"
            class="with-text"
            fill="currentColor"
            height="1em"
            role="img"
            style="vertical-align: -0.125em;"
            viewBox="0 0 448 512"
            width="1em"
          >
            <path
              d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34zm192-34l-136-136c-9.4-9.4-24.6-9.4-33.9 0l-22.6 22.6c-9.4 9.4-9.4 24.6 0 33.9l96.4 96.4-96.4 96.4c-9.4 9.4-9.4 24.6 0 33.9l22.6 22.6c9.4 9.4 24.6 9.4 33.9 0l136-136c9.4-9.2 9.4-24.4 0-33.8z"
              transform=""
            />
          </svg>
           
          <span
            title="src/main/java/org/your/GreetingApi.java"
          >
            greetingEndpoint
          </span>
        </div>
        <div
          class="http-request method-get"
        >
          <div
            class="definition"
          >
            <div
              class="http-request-def"
            >
              <span
                class="http-request-method"
              >
                GET
              </span>
               
              <span
                class="http-request-path"
              >
                /api/greeting?name=
              </span>
              <input
                aria-invalid="false"
                aria-label="Greetings name input"
                class="pf-c-form-control http-request-param"
                id="http-api-param-name-input"
                name="name"
                placeholder="World"
                type="text"
                value="John"
              />
              <button
                class="pf-c-button pf-m-primary copy-button"
                title="Copy as curl"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  fill="currentColor"
                  height="1em"
                  role="img"
                  style="vertical-align: -0.125em;"
                  viewBox="0 0 384 512"
                  width="1em"
                >
                  <path
                    d="M384 112v352c0 26.51-21.49 48-48 48H48c-26.51 0-48-21.49-48-48V112c0-26.51 21.49-48 48-48h80c0-35.29 28.71-64 64-64s64 28.71 64 64h80c26.51 0 48 21.49 48 48zM192 40c-13.255 0-24 10.745-24 24s10.745 24 24 24 24-10.745 24-24-10.745-24-24-24m96 114v-20a6 6 0 0 0-6-6H102a6 6 0 0 0-6 6v20a6 6 0 0 0 6 6h180a6 6 0 0 0 6-6z"
                    transform=""
                  />
                </svg>
              </button>
            </div>
          </div>
          <div
            class="action"
          >
            <button
              aria-label="Execute GET Greetings"
              class="pf-c-button pf-m-primary http-request-button"
              title="Execute GET Greetings"
              type="button"
            >
              Execute
            </button>
          </div>
        </div>
        <div
          class="pf-l-grid__item pf-m-12-col"
        >
          <div
            aria-label="REST Console"
            class="console"
          >
            <div
              class="console-content expanded"
            >
              <span
                class="placeholder"
              >
                <span
                  class="prefix"
                >
                  &gt;_
                </span>
                  Use the 'Execute' button(s) above to get things started in this console...
              </span>
              <div
                class="console-content-row"
              >
                <div>
                  <span
                    class="prefix"
                  >
                    $
                  </span>
                   
                  <span
                    class="time"
                  >
                    8:33:13 PM
                  </span>
                   
                  <span
                    class="method method-get"
                  >
                    GET
                  </span>
                   
                  <span
                    class="url"
                  >
                    http://mocked.io/api/greeting?name=John
                  </span>
                  :
                </div>
                <div
                  aria-label="Hello John!"
                >
                  Hello John!
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </article>
</DocumentFragment>
`;

exports[`<RestCapability /> check that initial render is correct 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <article
        class="pf-c-card capability-card"
        id="rest-capability"
      >
        <div
          class="pf-c-card__header capability-card-header"
        >
          <svg
            aria-hidden="true"
            class="with-text"
            fill="currentColor"
            height="1em"
            role="img"
            style="vertical-align: -0.125em;"
            viewBox="0 0 384 512"
            width="1em"
          >
            <path
              d="M256 144V32c0-17.673 14.327-32 32-32s32 14.327 32 32v112h-64zm112 16H16c-8.837 0-16 7.163-16 16v32c0 8.837 7.163 16 16 16h16v32c0 77.406 54.969 141.971 128 156.796V512h64v-99.204c73.031-14.825 128-79.39 128-156.796v-32h16c8.837 0 16-7.163 16-16v-32c0-8.837-7.163-16-16-16zm-240-16V32c0-17.673-14.327-32-32-32S64 14.327 64 32v112h64z"
              transform=""
            />
          </svg>
           
          HTTP APIs
        </div>
        <div
          class="pf-c-card__body capability-card-body"
        >
          <div
            class="pf-l-grid"
          >
            <div
              class="pf-l-grid__item pf-m-12-col"
            >
              HTTP API endpoints expose your application to outside callers. Through these, programs may communicate over the network in a language-independent fashion. We have created an initial set of endpoints to illustrate how you may accomplish this in your selected runtime. By composing together HTTP endpoints and making use of hypermedia and links, you may follow these patterns to construct a RESTful architecture.
            </div>
            <div
              class="capability-card-separator"
            >
              <hr />
            </div>
            <div
              class="pf-l-grid__item pf-m-12-col http-request-service"
            >
              <svg
                aria-hidden="true"
                class="with-text"
                fill="currentColor"
                height="1em"
                role="img"
                style="vertical-align: -0.125em;"
                viewBox="0 0 448 512"
                width="1em"
              >
                <path
                  d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34zm192-34l-136-136c-9.4-9.4-24.6-9.4-33.9 0l-22.6 22.6c-9.4 9.4-9.4 24.6 0 33.9l96.4 96.4-96.4 96.4c-9.4 9.4-9.4 24.6 0 33.9l22.6 22.6c9.4 9.4 24.6 9.4 33.9 0l136-136c9.4-9.2 9.4-24.4 0-33.8z"
                  transform=""
                />
              </svg>
               
              <span
                title="src/main/java/org/your/GreetingApi.java"
              >
                greetingEndpoint
              </span>
            </div>
            <div
              class="http-request method-get"
            >
              <div
                class="definition"
              >
                <div
                  class="http-request-def"
                >
                  <span
                    class="http-request-method"
                  >
                    GET
                  </span>
                   
                  <span
                    class="http-request-path"
                  >
                    /api/greeting?name=
                  </span>
                  <input
                    aria-invalid="false"
                    aria-label="Greetings name input"
                    class="pf-c-form-control http-request-param"
                    id="http-api-param-name-input"
                    name="name"
                    placeholder="World"
                    type="text"
                    value=""
                  />
                  <button
                    class="pf-c-button pf-m-primary copy-button"
                    title="Copy as curl"
                    type="button"
                  >
                    <svg
                      aria-hidden="true"
                      fill="currentColor"
                      height="1em"
                      role="img"
                      style="vertical-align: -0.125em;"
                      viewBox="0 0 384 512"
                      width="1em"
                    >
                      <path
                        d="M384 112v352c0 26.51-21.49 48-48 48H48c-26.51 0-48-21.49-48-48V112c0-26.51 21.49-48 48-48h80c0-35.29 28.71-64 64-64s64 28.71 64 64h80c26.51 0 48 21.49 48 48zM192 40c-13.255 0-24 10.745-24 24s10.745 24 24 24 24-10.745 24-24-10.745-24-24-24m96 114v-20a6 6 0 0 0-6-6H102a6 6 0 0 0-6 6v20a6 6 0 0 0 6 6h180a6 6 0 0 0 6-6z"
                        transform=""
                      />
                    </svg>
                  </button>
                </div>
              </div>
              <div
                class="action"
              >
                <button
                  aria-label="Execute GET Greetings"
                  class="pf-c-button pf-m-primary http-request-button"
                  title="Execute GET Greetings"
                  type="button"
                >
                  Execute
                </button>
              </div>
            </div>
            <div
              class="pf-l-grid__item pf-m-12-col"
            >
              <div
                aria-label="REST Console"
                class="console"
              >
                <div
                  class="console-content no-content"
                >
                  <span
                    class="placeholder"
                  >
                    <span
                      class="prefix"
                    >
                      &gt;_
                    </span>
                      Use the 'Execute' button(s) above to get things started in this console...
                  </span>
                </div>
              </div>
            </div>
          </div>
        </div>
      </article>
    </div>
  </body>,
  "container": <div>
    <article
      class="pf-c-card capability-card"
      id="rest-capability"
    >
      <div
        class="pf-c-card__header capability-card-header"
      >
        <svg
          aria-hidden="true"
          class="with-text"
          fill="currentColor"
          height="1em"
          role="img"
          style="vertical-align: -0.125em;"
          viewBox="0 0 384 512"
          width="1em"
        >
          <path
            d="M256 144V32c0-17.673 14.327-32 32-32s32 14.327 32 32v112h-64zm112 16H16c-8.837 0-16 7.163-16 16v32c0 8.837 7.163 16 16 16h16v32c0 77.406 54.969 141.971 128 156.796V512h64v-99.204c73.031-14.825 128-79.39 128-156.796v-32h16c8.837 0 16-7.163 16-16v-32c0-8.837-7.163-16-16-16zm-240-16V32c0-17.673-14.327-32-32-32S64 14.327 64 32v112h64z"
            transform=""
          />
        </svg>
         
        HTTP APIs
      </div>
      <div
        class="pf-c-card__body capability-card-body"
      >
        <div
          class="pf-l-grid"
        >
          <div
            class="pf-l-grid__item pf-m-12-col"
          >
            HTTP API endpoints expose your application to outside callers. Through these, programs may communicate over the network in a language-independent fashion. We have created an initial set of endpoints to illustrate how you may accomplish this in your selected runtime. By composing together HTTP endpoints and making use of hypermedia and links, you may follow these patterns to construct a RESTful architecture.
          </div>
          <div
            class="capability-card-separator"
          >
            <hr />
          </div>
          <div
            class="pf-l-grid__item pf-m-12-col http-request-service"
          >
            <svg
              aria-hidden="true"
              class="with-text"
              fill="currentColor"
              height="1em"
              role="img"
              style="vertical-align: -0.125em;"
              viewBox="0 0 448 512"
              width="1em"
            >
              <path
                d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34zm192-34l-136-136c-9.4-9.4-24.6-9.4-33.9 0l-22.6 22.6c-9.4 9.4-9.4 24.6 0 33.9l96.4 96.4-96.4 96.4c-9.4 9.4-9.4 24.6 0 33.9l22.6 22.6c9.4 9.4 24.6 9.4 33.9 0l136-136c9.4-9.2 9.4-24.4 0-33.8z"
                transform=""
              />
            </svg>
             
            <span
              title="src/main/java/org/your/GreetingApi.java"
            >
              greetingEndpoint
            </span>
          </div>
          <div
            class="http-request method-get"
          >
            <div
              class="definition"
            >
              <div
                class="http-request-def"
              >
                <span
                  class="http-request-method"
                >
                  GET
                </span>
                 
                <span
                  class="http-request-path"
                >
                  /api/greeting?name=
                </span>
                <input
                  aria-invalid="false"
                  aria-label="Greetings name input"
                  class="pf-c-form-control http-request-param"
                  id="http-api-param-name-input"
                  name="name"
                  placeholder="World"
                  type="text"
                  value=""
                />
                <button
                  class="pf-c-button pf-m-primary copy-button"
                  title="Copy as curl"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    fill="currentColor"
                    height="1em"
                    role="img"
                    style="vertical-align: -0.125em;"
                    viewBox="0 0 384 512"
                    width="1em"
                  >
                    <path
                      d="M384 112v352c0 26.51-21.49 48-48 48H48c-26.51 0-48-21.49-48-48V112c0-26.51 21.49-48 48-48h80c0-35.29 28.71-64 64-64s64 28.71 64 64h80c26.51 0 48 21.49 48 48zM192 40c-13.255 0-24 10.745-24 24s10.745 24 24 24 24-10.745 24-24-10.745-24-24-24m96 114v-20a6 6 0 0 0-6-6H102a6 6 0 0 0-6 6v20a6 6 0 0 0 6 6h180a6 6 0 0 0 6-6z"
                      transform=""
                    />
                  </svg>
                </button>
              </div>
            </div>
            <div
              class="action"
            >
              <button
                aria-label="Execute GET Greetings"
                class="pf-c-button pf-m-primary http-request-button"
                title="Execute GET Greetings"
                type="button"
              >
                Execute
              </button>
            </div>
          </div>
          <div
            class="pf-l-grid__item pf-m-12-col"
          >
            <div
              aria-label="REST Console"
              class="console"
            >
              <div
                class="console-content no-content"
              >
                <span
                  class="placeholder"
                >
                  <span
                    class="prefix"
                  >
                    &gt;_
                  </span>
                    Use the 'Execute' button(s) above to get things started in this console...
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
    </article>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllBySelectText": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getAllByValue": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getBySelectText": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "getByValue": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllBySelectText": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryAllByValue": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryBySelectText": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "queryByValue": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
